controller/src/controller.rs

```diff
@@ -319,6 +348,122 @@ impl<T: BottlerocketShadowClient> BrupopController<T> {
+            // This logic will cooperate with 6 mins buffer strategy.
+            if within_time_window(update_window_start, update_window_stop)? {
+                if !active_set.is_empty() {
+                    self.progress_active_set(active_set).await?;
+                }
+
+                // Bring one more node each time if the active nodes size is less than MAX_CONCURRENT_UPDATE setting.
+                if active_set_size < get_max_concurrent_update()? {
+                    // If there's nothing to operate on, check to see if any other nodes are ready for action.
+                    let new_active_node = self.find_and_update_ready_brs().await?;
+                    if let Some(brs) = new_active_node {
+                        event!(Level::INFO, name = %brs.name_any(), "Began updating new node.")
+                    }
+                }
+            } else {
+                // try to find out if any nodes are being updated. If yes, controller
+                // will complete them and pause actions on other waitingForUpdate nodes.
+                if !active_set.is_empty() {
+                    self.progress_active_set(active_set).await?;
+                } else {
+                    sleep(ACTION_INTERVAL).await;
+                    continue;
+                }
+            }
+
+            // Cleanup BRS when the operator is removed from a node
+            let brss_name = self
+                .all_brss()
+                .into_iter()
+                .map(|brs| brs.name_any())
+                .collect();
+            let nodes = self.all_nodes();
+            self.bottlerocketshadows_cleanup(nodes, brss_name).await?;
+
+            // Emit metrics at the end of the loop in case the loop didn't progress any nodes.
+            self.emit_metrics()?;
+
+            // Sleep until it's time to check for more action.
+            sleep(ACTION_INTERVAL).await;
+        }
+    }
+
+    /// Runs the event loop for the Brupop controller.
+    ///
+    /// Because the controller wants to gate the number of simultaneously updating nodes, we can't allow the update state machine
+    /// of each individual bottlerocket node to run concurrently and in an event-driven fashion, as is typically done with controllers.
+    /// Instead, we will keep an updated store of `BottlerocketShadow` objects based on cluster events, and then periodically make
+    /// scheduling decisions based on that store.
+    ///
+    /// The controller is designed to run on a single node in the cluster and rely on the scheduler to ensure there is always one
+    /// running; however, it could be expanded using leader-election and multiple nodes if the scheduler proves to be problematic.
+    pub async fn run(&self) -> Result<()> {
+        // generate brupop cron expression schedule
+        let brupop_schedule = generate_schedule()
+            .await
+            .context(controllerclient_error::GetCronScheduleSnafu)?;
```

Comment:
When I provided an invalid cron expression:
```yaml
- name: SCHEDULER_CRON_HOURS
  value: '99'
```
I get a crashloop backoff from the controller with unhelpful log errors:

```
❯ k logs -n brupop-bottlerocket-aws pod/brupop-controller-deployment-6cdcffd5d7-r5gh5
  2023-01-24T23:28:06.005178Z  INFO actix_server::builder: Starting 1 workers
    at /src/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-server-2.1.1/src/builder.rs:200

  2023-01-24T23:28:06.005276Z ERROR controller: controller exited
    at controller/src/main.rs:110
```

I'd expect an invalid cron expression to not error out like this and that we could catch some errors to inform the user that their cron expression is bad.

Reply:
I'll make it catch invalid input and exit with meaningful logs.